File: scripts/shared/manage_gitignore.py
#!/usr/bin/env python3
"""Maintain the managed ``.gitignore`` block in a downstream repo.

The workflow surface (skills, agents, docs, helper scripts) is committed to each
downstream repo's own git, so it travels with feature branches and
``git worktree`` checkouts. Machine-local runtime artifacts, plus the per-session
plan files under ``docs/plans/`` (local scratch the planner/executor writes per
bead — not part of the committed workflow surface), stay ignored. This module
writes that small managed block between the BEGIN/END markers and is idempotent:
re-running it replaces the previous block in place.
"""

from __future__ import annotations

import argparse
import json
import subprocess
from pathlib import Path


IGNORE_BLOCK_START = "# BEGIN TEMPLATE AGENT WORKFLOW LOCAL-ONLY"
IGNORE_BLOCK_END = "# END TEMPLATE AGENT WORKFLOW LOCAL-ONLY"
LEGACY_IGNORE_HEADER = "# Local agent workflow assets"

# Paths that must never be committed downstream: machine-local runtime artifacts,
# plus the per-session plan files under docs/plans/ (local scratch the
# planner/executor writes per bead — not part of the committed workflow surface).
# Everything else the template scaffolds is tracked in git. Older template
# versions listed every skill/agent/doc path here; ensure_ignore_block() strips
# the previous block before rewriting, so those stale entries (and any standalone
# docs/plans/ line a downstream added by hand) are folded into this block on the
# next run.
IGNORE_ENTRIES = (
    ".beads-credential-key",
    ".beads/interactions.jsonl",
    ".bv/",
    ".dolt/",
    "*.db",
    "scripts/shared/__pycache__/",
    "docs/plans/",
)


def resolve_repo_root(path: str | Path) -> Path:
    candidate = Path(path).expanduser().resolve()
    try:
        result = subprocess.run(
            ["git", "-C", str(candidate), "rev-parse", "--show-toplevel"],
            check=True,
            text=True,
            capture_output=True,
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        return candidate
    return Path(result.stdout.strip()).resolve()


def _strip_ignore_block(content: str) -> str:
    start = content.find(IGNORE_BLOCK_START)
    end = content.find(IGNORE_BLOCK_END)
    if start == -1 or end == -1 or end < start:
        return content
    end += len(IGNORE_BLOCK_END)
    before = content[:start].rstrip("\r\n")
    after = content[end:].lstrip("\r\n")
    if before and after:
        return before + "\n\n" + after
    return before or after


def _squash_blank_runs(lines: list[str]) -> list[str]:
    squashed: list[str] = []
    blank = False
    for line in lines:
        if line.strip():
            squashed.append(line.rstrip())
            blank = False
            continue
        if not blank:
            squashed.append("")
        blank = True
    while squashed and squashed[0] == "":
        squashed.pop(0)
    while squashed and squashed[-1] == "":
        squashed.pop()
    return squashed


def ensure_ignore_block(repo_root: Path) -> bool:
    target = repo_root / ".gitignore"
    content = target.read_text(encoding="utf-8") if target.exists() else ""
    stripped = _strip_ignore_block(content)

    filtered_lines: list[str] = []
    managed = set(IGNORE_ENTRIES)
    for line in stripped.splitlines():
        raw = line.rstrip("\r")
        if raw == LEGACY_IGNORE_HEADER:
            continue
        if raw in managed:
            continue
        filtered_lines.append(raw)
    filtered_lines = _squash_blank_runs(filtered_lines)

    block_lines = [
        IGNORE_BLOCK_START,
        *IGNORE_ENTRIES,
        IGNORE_BLOCK_END,
    ]

    merged = "\n".join(filtered_lines + ([""] if filtered_lines else []) + block_lines) + "\n"
    if merged == content:
        return False
    target.write_text(merged, encoding="utf-8")
    return True


def main() -> int:
    parser = argparse.ArgumentParser(description="Manage the downstream .gitignore workflow block")
    subparsers = parser.add_subparsers(dest="command", required=True)

    ensure_ignore = subparsers.add_parser(
        "ensure-ignore",
        help="Install or refresh the managed workflow ignore block in the downstream repo .gitignore",
    )
    ensure_ignore.add_argument("--repo", default=".", help="Downstream repo path (default: current directory)")
    ensure_ignore.add_argument("--json", action="store_true")

    args = parser.parse_args()
    if args.command == "ensure-ignore":
        repo_root = resolve_repo_root(args.repo)
        changed = ensure_ignore_block(repo_root)
        if args.json:
            print(json.dumps({"repo_root": str(repo_root), "changed": changed}))
        else:
            state = "updated" if changed else "already current"
            print(f"Managed workflow ignore block {state} in {repo_root / '.gitignore'}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: lib/manageGitignore.js
'use strict';

// Port of scripts/shared/manage_gitignore.py — maintain the managed `.gitignore`
// block in a downstream repo. Only machine-local runtime artifacts plus the
// per-session plan files under docs/plans/ are ignored; the rest of the workflow
// surface is committed downstream. Idempotent: re-running replaces the block.

const fs = require('fs');
const os = require('os');
const path = require('path');
const { runCapture } = require('./proc');

const IGNORE_BLOCK_START = '# BEGIN TEMPLATE AGENT WORKFLOW LOCAL-ONLY';
const IGNORE_BLOCK_END = '# END TEMPLATE AGENT WORKFLOW LOCAL-ONLY';
// Orphaned comment headers to drop from the downstream .gitignore. bd init writes
// its own "# Beads / Dolt files (added by bd init)" header + entries; our managed
// block already covers those entries, so the header is left stranded — strip it.
const STRIP_HEADERS = ['# Local agent workflow assets', '# Beads / Dolt files (added by bd init)'];

const IGNORE_ENTRIES = [
  '.beads-credential-key',
  '.beads/interactions.jsonl',
  '.beads/epic-runs',
  '.claude/scheduled_tasks.lock',
  '.claude/settings.local.json',
  '.bv/',
  '*.db',
  '.dolt/',
  'docs/plans/',
];

function expandHome(p) {
  if (p === '~') return os.homedir();
  if (p.startsWith('~/') || p.startsWith('~\\')) return path.join(os.homedir(), p.slice(2));
  return p;
}

function resolveRepoRoot(p) {
  const candidate = path.resolve(expandHome(String(p)));
  const res = runCapture('git', ['-C', candidate, 'rev-parse', '--show-toplevel']);
  if (res.status !== 0) {
    return candidate;
  }
  return path.resolve(res.stdout.trim());
}

// Emulate Python str.splitlines(): split on line boundaries, dropping a single
// trailing empty element produced by a final newline.
function splitlines(s) {
  const parts = s.split(/\r\n|\r|\n/);
  if (parts.length > 0 && parts[parts.length - 1] === '') {
    parts.pop();
  }
  return parts;
}

function stripIgnoreBlock(content) {
  const start = content.indexOf(IGNORE_BLOCK_START);
  const end = content.indexOf(IGNORE_BLOCK_END);
  if (start === -1 || end === -1 || end < start) {
    return content;
  }
  const endPos = end + IGNORE_BLOCK_END.length;
  const before = content.slice(0, start).replace(/[\r\n]+$/, '');
  const after = content.slice(endPos).replace(/^[\r\n]+/, '');
  if (before && after) {
    return before + '\n\n' + after;
  }
  return before || after;
}

function squashBlankRuns(lines) {
  const squashed = [];
  let blank = false;
  for (const line of lines) {
    if (line.trim()) {
      squashed.push(line.replace(/\s+$/, ''));
      blank = false;
      continue;
    }
    if (!blank) {
      squashed.push('');
    }
    blank = true;
  }
  while (squashed.length && squashed[0] === '') {
    squashed.shift();
  }
  while (squashed.length && squashed[squashed.length - 1] === '') {
    squashed.pop();
  }
  return squashed;
}

function ensureIgnoreBlock(repoRoot) {
  const target = path.join(repoRoot, '.gitignore');
  const content = fs.existsSync(target) ? fs.readFileSync(target, 'utf8') : '';
  const stripped = stripIgnoreBlock(content);

  const managed = new Set(IGNORE_ENTRIES);
  let filtered = [];
  for (const line of splitlines(stripped)) {
    const raw = line.replace(/\r+$/, '');
    if (STRIP_HEADERS.includes(raw)) continue;
    if (managed.has(raw)) continue;
    filtered.push(raw);
  }
  filtered = squashBlankRuns(filtered);

  const blockLines = [IGNORE_BLOCK_START, ...IGNORE_ENTRIES, IGNORE_BLOCK_END];
  const merged = [...filtered, ...(filtered.length ? [''] : []), ...blockLines].join('\n') + '\n';

  if (merged === content) {
    return false;
  }
  fs.writeFileSync(target, merged, 'utf8');
  return true;
}

// Entry point mirroring `manage_gitignore.py ensure-ignore --repo <path>`.
function ensureIgnore(repoPath) {
  const repoRoot = resolveRepoRoot(repoPath);
  const changed = ensureIgnoreBlock(repoRoot);
  return { repoRoot, changed };
}

module.exports = {
  ensureIgnore,
  ensureIgnoreBlock,
  resolveRepoRoot,
  IGNORE_ENTRIES,
  IGNORE_BLOCK_START,
  IGNORE_BLOCK_END,
};
